src/java.desktop/macosx/native/libosxapp/JNIUtilities.m

```diff
@@ -40,11 +40,18 @@
 }
 
 jstring NSStringToJavaString(JNIEnv* env, NSString *str) {
-
     if (str == NULL) {
        return NULL;
     }
-    jstring jStr = (*env)->NewStringUTF(env, [str UTF8String]);
+    jsize len = [str length];
+    unichar *buffer = (unichar*)calloc(len, sizeof(unichar));
+    if (buffer == NULL) {
+       return NULL;
+    }
+    NSRange crange = NSMakeRange(0, len);
+    [str getCharacters:buffer range:crange];
+    jstring jStr = (*env)->NewString(env, buffer, len);
+    free(buffer);
     CHECK_EXCEPTION();
```

Comment:
Shouldn't it be better to call CHECK_EXCEPTION_NULL_RETURN() here to return null in case NewString throw exception in which case, I guess jStr will have garbage.

Also, it seems https://developer.apple.com/documentation/foundation/nsstring/1408720-getcharacters says the buffer must be large enough...aRange.length*sizeof(unichar) so shouldn't the NSRange be created for 
(0, len*sizeof(unichar))

Reply:
>> Shouldn't it be better to call CHECK_EXCEPTION_NULL_RETURN()

If the jstring were garbage in such a case we would NOT want to check it against null, since
that would be a definite problem.

As it is NewString is specified to return NULL if there is a problem so the existing code is fine although the alternative pattern of CHECK_EXCEPTION_NULL_RETURN(jstr, NULL) would also work.

SInce I'm not changing this I think it fine to stick with the existing code.

>> the buffer must be large enough...aRange.length*sizeof(unichar) so shouldn't the NSRange be created for
(0, len * sizeof(unichar))

The buffer *is* large enough

buffer = (unichar*) calloc(len, sizeof(unichar));

the NSRange is the *number of elements* to copy. It is not the buffer storage